### PR TITLE
Expose metrics endpoint + Basic Int Tests

### PR DESCRIPTION
**Pull request by MiaAltieri**
Jira: [DPE-1893](https://warthogs.atlassian.net/browse/DPE-1893)

## Problem
Metrics endpoint not exposed for MongoDB 

## Solution
Expose metrics endpoint similarly to how the VM charm does 

## Int Tests
- Sanity check that endpoints work
- Check endpoints are still functioning after a network cut

## Future PRs
- Add functionality for changing the monitor user password and updating the endpoint correctly so that metrics are still provided

[DPE-1893]: https://warthogs.atlassian.net/browse/DPE-1893?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by delgod — APPROVED** [review]

File: src/charm.py [hunk 1]
```diff
@@ -326,7 +355,6 @@ def _pull_licenses(self, container: Container) -> None:
         """Pull licenses from workload."""
         licenses = [
             "snap",
-            "rock",
```
Comment (dmitry-ratushnyy) [inline src/charm.py:329]: Minor question - looks like this change is not related to the PR topic. Is this correct?

File: tests/integration/metrics_tests/test_metrics.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+# See LICENSE file for licensing details.
+import time
+import yaml
+
+import ops
+import pytest
+import urllib3
+from pytest_operator.plugin import OpsTest
+from pathlib import Path
+from ..ha_tests import helpers as ha_helpers
```
Comment (dmitry-ratushnyy) [inline tests/integration/metrics_tests/test_metrics.py:12]: is it possible to avoid relative imports?

File: tests/integration/metrics_tests/test_metrics.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+async def test_endpoints_network_cut(ops_test: OpsTest, chaos_mesh):
+    """Verify that endpoint still function correctly after a network cut."""
+    # retrieve a primary unit and a non-primary unit (active-unit). The primary unit will have its
+    # network disrupted, while the active unit allows us to communicate to `mongod`
+    primary = await ha_helpers.get_replica_set_primary(ops_test)
+    active_unit = [
+        unit
+        for unit in ops_test.model.applications[DATABASE_APP_NAME].units
+        if unit.name != primary.name
```
Comment (dmitry-ratushnyy) [inline tests/integration/metrics_tests/test_metrics.py:78]: maybe we can use `filter` here?

File: tests/integration/metrics_tests/test_metrics.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+@pytest.fixture(scope="module")
+def chaos_mesh(ops_test: OpsTest) -> None:
+    ha_helpers.deploy_chaos_mesh(ops_test.model.info.name)
+
+    yield
+
+    ha_helpers.destroy_chaos_mesh(ops_test.model.info.name)
+
+
+async def get_address(ops_test: OpsTest, app_name=DATABASE_APP_NAME, unit_num=0) -> str:
```
Comment (dmitry-ratushnyy) [inline tests/integration/metrics_tests/test_metrics.py:30]: It looks like `app_name` is always equals to `DATABASE_APP_NAME` if so, do we really need it as a param or am I missing something?

File: tests/integration/metrics_tests/test_metrics.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    address = status["applications"][app_name]["units"][f"{app_name}/{unit_num}"]["address"]
+    return address
+
+
+async def verify_endpoints(ops_test: OpsTest, unit: ops.model.Unit) -> str:
+    """Verifies mongodb endpoint is functional on a given unit."""
+    http = urllib3.PoolManager()
+
+    unit_address = await get_address(ops_test=ops_test)
+    mongodb_exporter_url = f"http://{unit_address}:{MONGODB_EXPORTER_PORT}/metrics"
```
Comment (dmitry-ratushnyy) [inline tests/integration/metrics_tests/test_metrics.py:42]: Maybe we can inline this variable?

File: tests/integration/metrics_tests/test_metrics.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        resources = {"mongodb-image": METADATA["resources"]["mongodb-image"]["upstream-source"]}
+        await ops_test.model.deploy(my_charm, num_units=3, resources=resources, series="jammy")
+        await ops_test.model.wait_for_idle(apps=[DATABASE_APP_NAME], status="active", timeout=2000)
+
+
+async def test_endpoints(ops_test: OpsTest):
+    """Sanity check that endpoints are running."""
+    application = ops_test.model.applications[DATABASE_APP_NAME]
+
+    for unit in application.units:
```
Comment (dmitry-ratushnyy) [inline tests/integration/metrics_tests/test_metrics.py:66]: Maybe we can use `asyncio.gather` here? :smile:

File: tests/integration/metrics_tests/test_metrics.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+async def verify_endpoints(ops_test: OpsTest, unit: ops.model.Unit) -> str:
+    """Verifies mongodb endpoint is functional on a given unit."""
+    http = urllib3.PoolManager()
+
+    unit_address = await get_address(ops_test=ops_test)
+    mongodb_exporter_url = f"http://{unit_address}:{MONGODB_EXPORTER_PORT}/metrics"
+    mongo_resp = http.request("GET", mongodb_exporter_url)
+
+    assert mongo_resp.status == 200
```
Comment (dmitry-ratushnyy) [inline tests/integration/metrics_tests/test_metrics.py:45]: Just curious why we don't use `assertEqual` from `unittest` here

File: tests/integration/metrics_tests/test_metrics.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    # if configured correctly there should be more than one mongodb metric present
+    mongodb_metrics = mongo_resp._body.decode("utf8")
+    assert mongodb_metrics.count("mongo") > 10
```
Comment (dmitry-ratushnyy) [inline tests/integration/metrics_tests/test_metrics.py:49]: Just curious why we don't use `assertGreater` from `unittest` here

File: src/charm.py (hunk 1 above)
Comment (MiaAltieri) [reply src/charm.py:329]: ooh this was a mistake - thank you for noticing 🙏

File: tests/integration/metrics_tests/test_metrics.py (hunk 4 above)
Comment (delgod) [reply tests/integration/metrics_tests/test_metrics.py:30]: It is needed it for one of the HA tests, which deploys two clusters in parallel and doesn't have mixed data.

File: tests/integration/metrics_tests/test_metrics.py (hunk 8 above)
Comment (MiaAltieri) [reply tests/integration/metrics_tests/test_metrics.py:49]: If I had to guess I would say because its an unnecessary import

File: tests/integration/metrics_tests/test_metrics.py (hunk 7 above)
Comment (MiaAltieri) [reply tests/integration/metrics_tests/test_metrics.py:45]: If I had to guess I would say because its an unnecessary import

File: tests/integration/metrics_tests/test_metrics.py (hunk 8 above)
Comment (dmitry-ratushnyy) [reply tests/integration/metrics_tests/test_metrics.py:49]: I think it is better to use `unitests` library when it is possible

File: tests/integration/metrics_tests/test_metrics.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+async def test_endpoints(ops_test: OpsTest):
+    """Sanity check that endpoints are running."""
+    application = ops_test.model.applications[DATABASE_APP_NAME]
+
+    for unit in application.units:
+        await verify_endpoints(ops_test, unit)
+
+
+async def test_endpoints_network_cut(ops_test: OpsTest, chaos_mesh):
```
Comment (dmitry-ratushnyy) [inline tests/integration/metrics_tests/test_metrics.py:70]: looks like `chaos_mesh` is not used in the test

File: tests/integration/metrics_tests/test_metrics.py (hunk 4 above)
Comment (delgod) [reply tests/integration/metrics_tests/test_metrics.py:30]: yes, @dmitry-ratushnyy is right, we don't use `get_address` function in the `test_unique_cluster_dbs` test.

File: tests/integration/metrics_tests/test_metrics.py (hunk 9 above)
Comment (delgod) [reply tests/integration/metrics_tests/test_metrics.py:70]: I think it is needed for `isolate_instance_from_cluster` and `remove_instance_isolation`

File: tests/integration/metrics_tests/test_metrics.py (hunk 6 above)
Comment (delgod) [reply tests/integration/metrics_tests/test_metrics.py:66]: I don't think it will add a significant speed increase, fetch URL should be super fast.

File: tests/integration/metrics_tests/test_metrics.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+async def test_build_and_deploy(ops_test: OpsTest) -> None:
+    """Build and deploy three units of MongoDB and one unit of TLS."""
+    async with ops_test.fast_forward():
+        my_charm = await ops_test.build_charm(".")
+        resources = {"mongodb-image": METADATA["resources"]["mongodb-image"]["upstream-source"]}
+        await ops_test.model.deploy(my_charm, num_units=3, resources=resources, series="jammy")
+        await ops_test.model.wait_for_idle(apps=[DATABASE_APP_NAME], status="active", timeout=2000)
+
+
+async def test_endpoints(ops_test: OpsTest):
```
[the remark this answers is not in the document]
Comment (delgod) [reply tests/integration/metrics_tests/test_metrics.py:62]: I think we can add mypy linter if we want strict typing.

File: tests/integration/metrics_tests/test_metrics.py (hunk 3 above)
Comment (delgod) [reply tests/integration/metrics_tests/test_metrics.py:78]: The generator (current) style has good readability (but inconsistent performance).
I think it is ok to keep it as is.

File: tests/integration/metrics_tests/test_metrics.py (hunk 9 above)
Comment (MiaAltieri) [reply tests/integration/metrics_tests/test_metrics.py:70]: yes this is correct!

File: tests/integration/metrics_tests/test_metrics.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply tests/integration/metrics_tests/test_metrics.py:62]: True - feel free to add `mypy` in a follow up PR Dmitry :)

File: tests/integration/metrics_tests/test_metrics.py (hunk 4 above)
Comment (MiaAltieri) [reply tests/integration/metrics_tests/test_metrics.py:30]: What we should be using is the HA helper:
```
mongodb_application_name = await ha_helpers.get_application_name(ops_test, DATABASE_APP_NAME)
```
which retrieves the app name so that we can run these tests on a provided cluster

File: tests/integration/metrics_tests/test_metrics.py (hunk 5 above)
Comment (MiaAltieri) [reply tests/integration/metrics_tests/test_metrics.py:42]: I think this would make things a little less readable. Is this a necessary change for you? If so I can make it

**Review by dmitry-ratushnyy — APPROVED** [review]
